### ISSUE TITLE
Cabal 3.12.0.1 reports that 'repl' does not support multiple targets

### ISSUE THREAD

**Issue by noughtmare** [haskell/cabal#10602]
**Describe the bug**

If I understand [this blog post](https://well-typed.com/blog/2024/07/hls-multi/) correctly, cabal 3.12.0.1 is supposed to have multi-component support. Unfortunately, trying to load multiple components with HLS still produces an error which contains this line:

```
The 'repl' command does not support multiple targets at once.
```

**To Reproduce**

```
$ git clone https://github.com/agda/agda.git
$ code agda
```

* open `src/full/Agda/Benchmarking.hs`
* wait for it to load
* open `test/Bugs/Tests.hs`
* see error

**Expected behavior**

No error.

**System information**
 - Operating system: NixOS
 - `cabal`, `ghc` versions:

```
$ cabal --version
cabal-install version 3.12.1.0
compiled using version 3.12.1.0 of the Cabal library 

$ ghc --version
The Glorious Glasgow Haskell Compilation System, version 9.6.6

$ haskell-language-server-wrapper --version
haskell-language-server version: 2.9.0.0 (GHC: 9.6.6) (PATH: /nix/store/w7h9j7ar2q1zk3dldh5adhbs85vm9kir-haskell-language-server-2.9.0.0/bin/haskell-language-server-wrapper)
```

**Comment by ulysses4ever** [comment]
Thanks for the report! Is there a reproducer without HLS and VS Code? If not, HLS repo may be a better place to start.

**Comment by noughtmare** [comment]
I can reproduce it with just cabal:

```
$ cabal repl --enable-multi-repl lib:agda test:agda-tests
Warning: The package list for 'hackage.haskell.org' is 39 days old.
Run 'cabal update' to get the latest list of available packages.
Resolving dependencies...
Build profile: -w ghc-9.6.6 -O1
In order, the following will be built (use -v for more details):
 - Agda-2.8.0 (interactive) (configuration changed)
[1 of 2] Compiling Main             ( /home/user/haskell/agda/dist-newstyle/build/x86_64-linux/ghc-9.6.6/Agda-2.8.0/setup/setup.hs, /home/user/haskell/agda/dist-newstyle/build/x86_64-linux/ghc-9.6.6/Agda-2.8.0/setup/Main.o ) [Distribution.Simple.BuildPaths package changed]
[2 of 2] Linking /home/user/haskell/agda/dist-newstyle/build/x86_64-linux/ghc-9.6.6/Agda-2.8.0/setup/setup [Objects changed]
Configuring Agda-2.8.0...
Error: [Cabal-6091]
The 'repl' command does not support multiple targets at once.

Error: [Cabal-7125]
repl failed for Agda-2.8.0.
```

**Comment by noughtmare** [comment]
It seems that this is specific to Agda. Running this command on a fresh cabal project with a library and test suite does work.

**Comment by ulysses4ever** [comment]
Oh, Agda is a wonderful source of Cabal bugs! Unfortunately, some of them has to do with the Custom build type, which is on life support these days... I wonder if there's a simpler reproducer with Custom...

**Comment by noughtmare** [comment]
You are right. I created a simple package with two components and a trivial custom-setup and it gives me the same error. I guess custom-setup is just not supported. Perhaps the best immediate course of action is to point that out in the error message.

**Comment by mpickering** [comment]
You need to build the `Setup` with a version of the cabal library which supports the multi-repl. I am not sure why this is allowed as there should be a constraint which forces a new enough version of `Cabal` to be used.

**Comment by noughtmare** [comment]
Is that something I could do myself? I tried adding this constraint in the `custom-setup` section:

```
custom-setup
  setup-depends:
    base, Cabal >=3.12
```

But that doesn't help.

**Comment by mpickering** [comment]
@andreasabel If you are interested then you can migrate your Custom setup to the new Hooks at some point in the future (a patch was prepared here - https://gitlab.haskell.org/mpickering/hooks-setup-testing/-/blob/master/patches/Agda-2.6.3.patch)

Supporting Agda's usecase was considered in the design process so hopefully that would work well for you.

**Comment by mpickering** [comment]
@andreasabel That's ok, it also sounds like a good design as well. It's interesting that you have reported that you have chosen to go down that route.

**Comment by georgefst** [comment]
Note that this also happens with `build-type: Hooks`. 

@sheaf tells me the core issue is #9986.

Also related: #10983, #10966.

**Comment by georgefst** [comment]
To be clear, this is only an issue when the package with the custom setup has multiple components, e.g. a Cabal file defining a library and an executable.

It is otherwise perfectly possible to use `--enable-multi-repl` on a project with multiple packages, one of which has a custom setup.
